### ISSUE TITLE
Jack sequencer seems to be broken on Fedora 38

### ISSUE THREAD

**Issue by ycollet** [helge17/tuxguitar#401]
I use tuxguitar 1.6.2 on Fedora 38 with pipewire-jack.
pipewire-jack is version 1.0.3.
I selected "jack sequencer" and a fluidsynth output on tuxguitar.
When I played a score, the sound is quite strange.

**Comment by ycollet** [comment]
The "realtime sequencer" and the "tuxguitar sequencer" work fine.

**Comment by guiv42** [comment]
Trying to reproduce in my environment (openSuse, TuxGuitar 1.6.2 installed from distro's repo): with all plugins enabled, if I set MIDI Sequencer to "Jack sequencer" I don't have any fluidsynth option available in MIDI port drop-down list

**Comment by ycollet** [comment]
I remember that on some distro, the fluidsynth plugin was in a tuxguitar-fluildsynth package ... Is this the case for you ?

**Comment by guiv42** [comment]
no, there's only one tuxguitar-related package in openSuse repository, the one I have packaged ;)

**Comment by ycollet** [comment]
I added a custom sf2 sound font in the fluidsynth plugin.
https://musical-artifacts.com/artifacts/362 -> Xiaod Bank Soundfont
Maybe you don't see the fluidsynth entry  because no default soundfont are provided

**Comment by guiv42** [comment]
> Maybe you don't see the fluidsynth entry because no default soundfont are provided

Yes, that's right, thanks.
Now if I want to test/reproduce, I need to setup a Jack configuration. I've never done that, and I don't know Jack at all.
Not much time available these days, so unless someone else can investigate, this issue will require time.

**Comment by helge17** [comment]
I tested TuxGuitar 1.6.3beta2 with Jack2 (not Pipewire-Jack!) and FluidSynth on my Debian 12.5 system and it works:

Start TuxGuitar with default settings

Go to Tools -> Plugins, activate Jack & FluidSynth:
[x] Jack Audio Connection Kit plugin support
[x] FluidSynth output plugin
    Configure Plugin: Audio -> Audio Driver: jack
    Keep default settings for the rest (Soundfont: ${tuxguitar.home.path}/share/soundfont/MagicSFver2.sf2 - may be different in the official RPM/DEB packages)
OK -> Close

Go to Tools -> Settings -> Sound:
MIDI Sequencer: Jack Sequencer
MIDI Port: TG Fluidsynth [MagicSFver2]
OK

TuxGuitar starts Jack in background, in my case ps shows the following process:
/usr/bin/jackd -T -ndefault -T -d alsa

Open file & play: Player starts, but no sound is coming out of the loudspeakers.

Start QjackCtl

Click on the Graph button, then in the Graph window:
- Select both "fluidsynth (left/right)" and "system (playback_1/playback_2)" with Ctrl-click.
- Click Connect

The TuxGuitar sound now comes out of the speakers loud and clear. You can also control the TuxGuitar playback with the player controls in QjackCtl.

Currently I am not able to test this with Pipewire-Jack, as switching my system to Pipewire would require too many changes and I already had problems with Pipewire in the past.

**Comment by ycollet** [comment]
Thanks a lot for this complete test.
I will redo the test on my side once 1.6.3 will be released.
In the mean time, all the other sequencers are working fine :)

**Comment by guiv42** [comment]
@ycollet : when you say Jack sequencer is *broken*, does it mean it was working correctly in your environment *before* 1.6.2?
If so, does it seem realistic for you to `git bisect` this issue, to identify which commit introduced this?
Unfortunately I cannot do it myself as I don't have an appropriate environment available to reproduce.

**Comment by guiv42** [comment]
I'm trying to clean up a bit the list of open issues.
@ycollet : what is the current status of this issue? Since the problem is not precisely characterized and we have no means to reproduce issue, I don't see how it could progress.

**Comment by ycollet** [comment]
Tested again on Fedora 39 with tuxguitar 1.6.4. I recorded a small example with the real time sequencer and the jack sequencer to show the problem. I will try to investigate a little more and will post the audio files today ...

**Comment by ycollet** [comment]
A file recorded with realtime sequencer (same result with tuxguitar sequencer):
https://soundcloud.com/josephcurwen-1/tuxguitar-normal?si=47ed1124c9754875b0feac93f8dbe8b8&utm_source=clipboard&utm_medium=text&utm_campaign=social_sharing

A file recorded with jack sequencer (the sound is totally distorted):
https://soundcloud.com/josephcurwen-1/tuxguitar-jacksequencer?si=47ed1124c9754875b0feac93f8dbe8b8&utm_source=clipboard&utm_medium=text&utm_campaign=social_sharing

**Comment by guiv42** [comment]
OK, so the question now is: what did change to provoke this problem? This could be either TuxGuitar or something in its environment.
Could you please try to identify the most recent version of TuxGuitar without this issue?

**Comment by ycollet** [comment]
I will try with older versions of tuxguitar and keep you informed

**Comment by guiv42** [comment]
Closing this issue (I assume it's now obsolete, Fedora 43 was released).
Please re-open it if needed.

**Comment by ycollet** [comment]
I still have the problem with TuxGuitar-2.0.0 and Fedora 43

**Comment by ycollet** [comment]
I made a video where the bug can be heard:
https://www.youtube.com/watch?v=qsmWLHUCrlo

**Comment by guiv42** [comment]
ok, thanks for the update

**Comment by guiv42** [comment]
@ycollet : could you please have a try at [this branch](https://github.com/guiv42/tuxguitar/tree/jackSequencer)?

To be honest I don't fully understand the consequences of this commit. However, the risk is low because the Jack sequencer is currently completely unusable. Modification is limited to `JackSequencer`, so practically I can't break anything ;)

I really don't understand the historic implementation (bug already present in historic version 1.5.6):
- each time the midi tick was updated, it called `JackSequencer.setTickPosition`
- this method was setting `this.reset` to `true`
- so, at next iteration, the sequencer called `this.reset()`

As a result, every 20ms the following methods were called: `sendAllNotesOff()` and `sendPitchBendReset()`
No wonder it did not work correctly

**Comment by ycollet** [comment]
I will test your fix next week. Thanks a lot for this work !

**Comment by ycollet** [comment]
Just tested the JackSequencer branch. I tested the Realtime / TuxGuitar and Jack sequencer and everything was fine.
So, it looks like the bug is fixed :)

**Comment by guiv42** [comment]
OK, thank you
fix is now merged
